fastchat/serve/openai_api_server.py

```diff
@@ -177,59 +212,165 @@ async def _get_worker_address(model_name: str, client: httpx.AsyncClient) -> str
+@app.get("/v1/models")
+async def show_available_models():
+    controller_address = app_settings.controller_address
     async with httpx.AsyncClient() as client:
-        worker_addr = await _get_worker_address(model_name, client)
-        delimiter = b"\0"
- 
-        for idx in range(n):
-            delta_position = 0
-            response = ChatCompletionStreamResponse(
-                model=model_name,
-                choices=[
-                    ChatCompletionResponseStreamChoice(
-                        index=idx,
-                        delta=DeltaMessage(content=""),
-                        finish_reason=None,
-                    )
-                ]
+        ret = await client.post(controller_address + "/refresh_all_workers")
+        ret = await client.post(controller_address + "/list_models")
+    models = ret.json()["models"]
+    models.sort()
+    return {"data": [{"id": m, "object": "model"} for m in models], "object": "list"}
+
+
+@app.post("/v1/chat/completions")
+async def create_chat_completion(request: ChatCompletionRequest):
+    """Creates a completion for the chat message"""
+    error_check_ret = await check_model(request)
+    if error_check_ret is not None:
+        return error_check_ret
+    error_check_ret = check_requests(request)
+    if error_check_ret is not None:
+        return error_check_ret
+
+    gen_params = get_gen_params(
+        request.model,
+        request.messages,
+        temperature=request.temperature,
+        top_p=request.top_p,
+        max_tokens=request.max_tokens,
+        echo=False,
+        stream=request.stream,
+        stop=request.stop,
+    )
+    if request.stream:
+        generator = chat_completion_stream_generator(request.model, gen_params, request.n)
+        return StreamingResponse(generator, media_type="text/event-stream")
+
+    choices = []
+    # TODO: batch the requests. maybe not necessary if using CacheFlow worker
+    chat_completions = []
+    for i in range(request.n):
+        content = asyncio.create_task(chat_completion(request.model, gen_params))
+        chat_completions.append(content)
+
+    usage = {
+        "prompt_tokens": 0,
+        "completion_tokens": 0,
+        "total_tokens": 0,
+    }
+    for i, content_task in enumerate(chat_completions):
+        try:
+            content = await content_task
+        except Exception as e:
+            return create_error_response(
+                50090,
+                str(e)
+            )
+        if content["error_code"] != 0:
+            return create_error_response(content["error_code"], content["text"])
+        choices.append(
+            ChatCompletionResponseChoice(
+                index=i,
+                message=ChatMessage(role="assistant", content=content["text"]),
+                finish_reason=content.get("finish_reason", "stop"),
+            )
+        )
+        for usage_k, usage_v in content["usage"].items():
```

Comment:
parse `content["usage"]` into `UsageInfo` first, to avoid manipulating python dict?